lib/token.js

```diff
@@ -464,53 +511,81 @@ function getWithoutPrompt(sdk, oauthOptions, options) {
-        // Default the responseMode to query
-        util.extend(oauthParams, {
-          responseMode: 'query'
-        });
-    // Otherwise, default to fragment
-    } else {
-      util.extend(oauthParams, {
-        responseMode: 'fragment'
-      });
-    }
+
+  if (oauthParams.grantType !== 'authorization_code') {
+    return Q.resolve(oauthParams);
   }
 
-  var urls = oauthUtil.getOAuthUrls(sdk, oauthParams, options);
-  var requestUrl = urls.authorizeUrl + buildAuthorizeParams(oauthParams);
+  return oauthUtil.getWellKnown(sdk, null)
+    .then(function(res) {
+      var methods = res['code_challenge_methods_supported'] || [];
+      if (methods.indexOf(oauthParams.codeChallengeMethod) === -1) {
+        throw new AuthSdkError('Invalid code_challenge_method');
+      }
+    })
+    .then(function() {
+      // PKCE authorization_code flow
+      var codeVerifier = pkce.generateVerifier(oauthParams.codeVerifier);
+
+      // We will need these values after redirect when we call /token
+      var meta = {
+        codeVerifier: codeVerifier,
+        responseType: oauthParams.responseType,
+        redirectUri: oauthParams.redirectUri
+      };
+      pkce.saveMeta(sdk, meta);
+
+      return pkce.computeChallenge(codeVerifier);
+    })
+    .then(function(codeChallenge) {
 
-  // Set session cookie to store the oauthParams
-  cookies.set(config.REDIRECT_OAUTH_PARAMS_COOKIE_NAME, JSON.stringify({
-    responseType: oauthParams.responseType,
-    state: oauthParams.state,
-    nonce: oauthParams.nonce,
-    scopes: oauthParams.scopes,
-    clientId: oauthParams.clientId,
-    urls: urls,
-    ignoreSignature: oauthParams.ignoreSignature
-  }));
+      // Clone/copy the params. Set codeChallenge and responseType for authorization_code
+      var clonedParams = util.clone(oauthParams) || {};
+      util.extend(clonedParams, oauthParams, {
+        codeChallenge: codeChallenge,
+        responseType: 'code'
+      });
+      return clonedParams;
+    });
+}
 
-  // Set nonce cookie for servers to validate nonce in id_token
-  cookies.set(config.REDIRECT_NONCE_COOKIE_NAME, oauthParams.nonce);
+function getWithRedirect(sdk, oauthOptions, options) {
+  oauthOptions = util.clone(oauthOptions) || {};
+  if (!oauthOptions.responseMode) {
+    oauthOptions.responseMode = 'fragment';
+  }
+  return prepareOauthParams(sdk, oauthOptions)
+    .then(function(oauthParams) {
+      var urls = oauthUtil.getOAuthUrls(sdk, oauthParams, options);
+      var requestUrl = urls.authorizeUrl + buildAuthorizeParams(oauthParams);
+
+      // Set session cookie to store the oauthParams
+      cookies.set(config.REDIRECT_OAUTH_PARAMS_COOKIE_NAME, JSON.stringify({
+        responseType: oauthParams.responseType,
+        state: oauthParams.state,
+        nonce: oauthParams.nonce,
+        scopes: oauthParams.scopes,
+        clientId: oauthParams.clientId,
+        urls: urls,
+        ignoreSignature: oauthParams.ignoreSignature
```

Comment:
Any reason we cannot use this cookie to store the PKCE values? This will allow us to remove the `load`, `save`, and `clear` meta methods.

Reply:
I do not want to risk the code verifier being transmitted over an insecure connection. We store PKCE in local storage, with fallback to session storage, then cookie (which I do not entirely support). There is an update on the way to make secure cookies (and https) the default, but it still bothers me that the codeVerifier and codeChallenge could be transmitted unencrypted. I might prefer to remove cookie fallback for PKCE support and add local/session storage support as a requirement in `isPKCESupported()`